src/services/env/Env.ts

```diff
@@ -1,48 +1,51 @@
 import { PATH_CONFIG_DEFAULT } from '@/pathConfigDefault'
 import { PathConfig } from '@/types/index'
 
-export type EnvVars = {
-  KUMA_NAME: string
+export type EnvArgs = {
+  KUMA_PRODUCT_NAME: string
+  KUMA_FEEDBACK_URL: string
+  KUMA_CHAT_URL: string
+  KUMA_INSTALL_URL: string
+  KUMA_DOCS_URL: string
+}
+type EnvProps = {
   KUMA_VERSION: string
   KUMA_BASE_PATH: string
   KUMA_API_URL: string
-  KUMA_DOCS_URL: string
   KUMA_UTM_QUERY_PARAMS: string
-  KUMA_FEEDBACK_URL: string
-  KUMA_CHAT_URL: string
-  KUMA_INSTALL_URL: string
 }
+export type EnvVars = EnvArgs & EnvProps
+
+type EnvInternal = EnvArgs & Partial<EnvProps>
 export default class Env {
   env: EnvVars
-  constructor(env: EnvVars) {
-    this.env = env
+  constructor(envArgs: EnvArgs) {
+    let _env: EnvInternal = envArgs
+    const env = (str: keyof EnvInternal, d: string = '') => this.var(str, this.var(str, _env?.[str] ?? d))
+
+    const config = readPathConfigFromDom()
+    const version = semver(env('KUMA_VERSION', config.version))
+
+    _env = {
+      ..._env,
+      KUMA_UTM_QUERY_PARAMS: `utm_source=${env('KUMA_PRODUCT_NAME')}&utm_medium=${env('KUMA_PRODUCT_NAME')}-GUI`,
+    }
+    this.env = {
+      ..._env as EnvVars,
+      KUMA_INSTALL_URL: `${env('KUMA_INSTALL_URL')}?${env('KUMA_UTM_QUERY_PARAMS')}`,
+      // TODO(jc): not totally sure we need to use a regex here, maybe just split and join if not
+      KUMA_DOCS_URL: `${env('KUMA_DOCS_URL')}/${version.patch === '0.0.0' ? 'dev' : version.patch.replace(/\.\d+$/, '.x')}`,
+      KUMA_VERSION: version.pre,
+      KUMA_API_URL: env('KUMA_API_URL') ?? config.apiUrl,
+      KUMA_BASE_PATH: env('KUMA_BASE_PATH') ?? config.baseGuiPath,
+    }
   }
 
-  var(key: keyof EnvVars) {
-    return this.env[key]
+  var(key: keyof EnvVars, d: string = '') {
+    return this.env?.[key] ?? d
```

Comment:
**Question**: I don’t think `this.env` can actually be `undefined` after construction. Can you please check this?

Reply:
On line 24:

```javascript
const env = (str: keyof EnvInternal, d: string = '') => this.var(str, this.var(str, _env?.[str] ?? d))
```
If you call `env` before you set `this.env` (on line 33) - which we do, then `this.env` is `undefined`. More to the point tho, I think there's one too many `this.var`s on line 24 😆

Reply:
I see, but does it even make sense to call `this.var` there? It can’t ever find a value because `this.env` will be `undefined` and so it would fall back to the default value.

Reply:
Yes, because if you override `var` so that that it checks elsewhere first before checking `this.env` (i.e. Cookies) then it will find a value.

Reply:
Hm. I see. It’s really quite intransparent to account for that in a utility that shouldn’t really have the outside knowledge.

In any case, the declaration of `env` in line 21 needs to be typed as `env: EnvVars | undefined` to account for this.

Reply:
Neither my text editor nor VSCode pokes me about this, I'm guessing because it is set in the constructor (which is quite common), i.e. after you've constructed an instance it _is_ defined. I added it here anyway even though as a consumer it is pretty much always set. In the scheme of things I don't mind either way.

Reply:
You can observe the issue with this by changing the implementation of `var` as follows:

```ts
  var(key: keyof EnvVars, d: string = '') {
    return this.env[key] ?? d
  }
```

It doesn’t result in a TypeScript error but it should (hence why optional chaining is needed) and that’s because the type of `this.env` is incomplete.